great_expectations/core/batch_definition.py

```diff
@@ -73,7 +74,18 @@ def get_batch(self, batch_parameters: Optional[BatchParameters] = None) -> Batch
         return batch_list[-1]
 
     def save(self) -> None:
-        self.data_asset._save_batch_definition(self)
+        from great_expectations.data_context.data_context.context_factory import project_manager
+
+        datasource = self.data_asset.datasource
+        datasource_dict = project_manager.get_datasources()
+
+        try:
+            datasource_dict[datasource.name] = datasource
```

Comment:
Does this work because we have a `DatasourceDict` which overrides `__setitem__`?
This seems strange since everywhere else we set things up in memory and then call save. But we can't change this now.
If we have access to the context, we could call `context.add_or_update_datasource` I believe.